examples/platform/esp32/route_hook/esp_route_table.c

```diff
@@ -0,0 +1,179 @@
+{
+    for (size_t i = 0; i < LWIP_ARRAYSIZE(s_route_entries); i++)
+    {
+        if (s_route_entries[i].netif == NULL)
+        {
+            break;
+        }
+        if (s_route_entries[i].netif == route_entry->netif && s_route_entries[i].prefix_length == route_entry->prefix_length &&
+            memcmp(s_route_entries[i].gateway.addr, route_entry->gateway.addr, sizeof(route_entry->gateway.addr)) == 0 &&
+            memcmp(s_route_entries[i].prefix.addr, route_entry->prefix.addr, route_entry->prefix_length / 8) == 0)
+        {
+            return &s_route_entries[i];
+        }
+    }
+    return NULL;
+}
+
+static esp_route_entry_t * find_empty_route_entry(void)
+{
+    for (size_t i = 0; i < LWIP_ARRAYSIZE(s_route_entries); i++)
+    {
+        if (s_route_entries[i].netif == NULL)
+        {
+            return &s_route_entries[i];
+        }
+    }
+    return NULL;
+}
+
+static void route_timeout_handler(void * arg)
+{
+    esp_route_entry_t * route = (esp_route_entry_t *) arg;
+
+    esp_route_table_remove_route_entry(route);
+}
+
+esp_route_entry_t * esp_route_table_add_route_entry(const esp_route_entry_t * route_entry)
+{
+    if (route_entry == NULL)
+    {
+        return NULL;
+    }
+
+    esp_route_entry_t * entry = find_route_entry(route_entry);
+
+    if (entry == NULL)
+    {
+        entry = find_empty_route_entry();
+        if (entry == NULL)
+        {
+            return NULL;
+        }
+        entry->netif   = route_entry->netif;
+        entry->gateway = route_entry->gateway;
+        ip6_addr_assign_zone(&entry->gateway, IP6_UNICAST, entry->netif);
+        entry->prefix        = route_entry->prefix;
+        entry->prefix_length = route_entry->prefix_length;
+    }
+    else
+    {
+        sys_untimeout(route_timeout_handler, entry);
+    }
+    entry->preference       = route_entry->preference;
+    entry->lifetime_seconds = route_entry->lifetime_seconds;
+    sys_timeout(entry->lifetime_seconds * 1000, route_timeout_handler, entry);
+    return entry;
+}
+
+esp_err_t esp_route_table_remove_route_entry(esp_route_entry_t * route_entry)
+{
+    if (route_entry < &s_route_entries[0] || route_entry > &s_route_entries[LWIP_ARRAYSIZE(s_route_entries)])
+    {
+        return ESP_ERR_INVALID_ARG;
+    }
+    route_entry->netif = NULL;
+    for (esp_route_entry_t * moved = route_entry; moved < &s_route_entries[LWIP_ARRAYSIZE(s_route_entries) - 1]; moved++)
+    {
+        *moved = *(moved + 1);
+        if (moved->netif == NULL)
+        {
+            break;
+        }
+    }
+    return ESP_OK;
+}
+
+static inline bool is_better_route(const esp_route_entry_t * lhs, const esp_route_entry_t * rhs)
+{
+    if (rhs == NULL)
+    {
+        return true;
+    }
+    if (lhs == NULL)
+    {
+        return false;
+    }
+    return lhs->preference > rhs->preference || (lhs->preference == rhs->preference && lhs->prefix_length > rhs->prefix_length);
```

Comment:
I think this might be backwards - longer length overrides preference. https://datatracker.ietf.org/doc/html/rfc4191 S. 3.2